### ISSUE TITLE
Support wasm (in some form)

### ISSUE THREAD

**Issue by Thomasdezeeuw** [rust-lang/socket2#268]
We can support wasm via [wasi](https://docs.rs/wasi). Last I looked (for https://github.com/tokio-rs/mio/pull/1395, back in November 2020) it doesn't support creating sockets, but it does support some things like read/write.

Alternatively we could start with an implementation that just panics, just so socket2 can build on wasm.

**Comment by ChenKS12138** [comment]
> We can support wasm via [wasi](https://docs.rs/wasi). Last I looked (for [tokio-rs/mio#1395](https://github.com/tokio-rs/mio/pull/1395), back in November 2020) it doesn't support creating sockets, but it does support some things like read/write.
> 
> Alternatively we could start with an implementation that just panics, just so socket2 can build on wasm.

I'm interested in using socket2 in wasm too. I found that symbols like AF_INET, sa_family_t, sockaddr_in are missing in [wasi](https://docs.rs/wasi), so it doesn't have the concept about the address of socket, and we barely could define `SockAddr` for wasm. [wasi](https://github.com/WebAssembly/WASI/blob/0ba0c5e2e37625ca5a6d3e4255a998dfaa3efc52/phases/snapshot/docs.md) supports socket_accept, socket_recv, socket_send, socket_shutdown by far. [accept](https://github.com/rust-lang/socket2/blob/c288e26816b06829e58a676ef4c5d2559be043b9/src/socket.rs#L256) in socket2 needs to return a `SockAddr` as the address of peer socket. So I think it seems hard to start with an implementation that just panics.

**Comment by sascha1337** [comment]
the day, wasi will have them features, will be gold for rust and thus the planet ... but wen X_X

**Comment by notgull** [comment]
Isn’t `wasi` an unstable API at the moment? It might not be the best idea to build on top of it.

**Comment by Thomasdezeeuw** [comment]
The API is in preview I believe, but considering it's implemented by browser backwards compatibility is everything. So I think it's find to add experimental support for wasm/wasi, this is what we did for Mio.

**Comment by thaodt** [comment]
> The API is in preview I believe, but considering it's implemented by browser backwards compatibility is everything. So I think it's find to add experimental support for wasm/wasi, this is what we did for Mio.

hi @Thomasdezeeuw do we have any ways to work around now? My lib depends on `tokio` and `tokio` depends on `socket2`. As a result, it built failed under target `wasm32-unknown-unknown` (`wasm-pack build --target web`)

**Comment by sascha1337** [comment]
Just do some draft PoC please, we need em so keta and tokio

**Comment by Thomasdezeeuw** [comment]
> hi @Thomasdezeeuw do we have any ways to work around now? My lib depends on `tokio` and `tokio` depends on `socket2`. As a result, it built failed under target `wasm32-unknown-unknown` (`wasm-pack build --target web`)

You might be able to disable Tokio's network feature, not sure though.

> Just do some draft PoC please, we need em sockets and tokio

Prs are welcome.

**Comment by Darksonn** [comment]
Tokio only depends on socket2 if you enable the `net` feature, and even if you make socket2 compile under wasm somehow, I'm pretty sure that there would also be other problems with getting Tokio's `net` feature to work.

**Comment by miketwenty1** [comment]
can verify trying to use either crate:
`awc = "3.0.0"`
`actix-rt = "2.7.0"`
Will result in a bunch of socket2 errors when trying to compile to wasm. 
such as:
```
/Users/tzez/.cargo/registry/src/github.com-1ecc6299db9ec823/socket2-0.4.4/src/socket.rs:1781:15
     |
1781 | from!(Socket, net::UdpSocket);
     |               ^^^^^^^^^^^^^^ expected struct `UdpSocket`, found `()`
```
Is it assumed until this issue is resolved that we won't be able to use socket2 with wasm or any libs/crates that rely on socket2?

**Comment by Lunarequest** [comment]
this should be easier to solve now that [wasix](https://wasix.org) exists, I'm not to sure how sockets work so I haven't taken more then a cursory build which resulting in more errors(yes you are reading that right, it seems a lot of the wasix sockets stuff expects more/less args)

**Comment by Thomasdezeeuw** [comment]
I don't know about wasix, it's doesn't seem like a standard anyone but wasmer is following. I don't really fancy having different code for different wasm runtimes.

**Comment by Lunarequest** [comment]
Wasix seems more like a pollyfill. It produces a wasm binary that should work any where

**Comment by Thomasdezeeuw** [comment]
I highly doubt that. I can't imagine how you would create a socket if the underlying wasm runtime doesn't provide a function/"system call" to create sockets.

**Comment by Lunarequest** [comment]
They are able to do this via a [npm package](https://www.npmjs.com/package/@wasmer/wasi) and have also [forked](https://github.com/wasix-org/socket2.git) this repo to allow it to be built to wasm. Having it upstreamed would be very cool

**Comment by Thomasdezeeuw** [comment]
Cool, a pr with those changes would be welcome. Ideally we can setup some CI as well, running `cargo check` is good enough, but of course `cargo test` would be preferred.

**Comment by gibbz00** [comment]
What changes does the WASIX fork have that aren't included in master? I can't seem to find any 🙈

**Comment by Thomasdezeeuw** [comment]
I'm ok with supporting WASI preview 2, but rustc isn't quite ready yet I think, see https://blog.rust-lang.org/2024/04/09/updates-to-rusts-wasi-targets.html (which you also linked).

I think WASIX is a no-go as that's vendor specific. Maybe as an alternative target, but I'm not how that would work at the moment.

Overall I'm also ok with reducing the socket2 to something that is supported on wasi preview 1 (and only wasi) so that it can at least build.

**Comment by LetsMelon** [comment]
> I'm ok with supporting WASI preview 2, but rustc isn't quite ready yet I think, see https://blog.rust-lang.org/2024/04/09/updates-to-rusts-wasi-targets.html (which you also linked).

Yeah, I think it's a bit too early for WASI preview 2. However, I believe that when the target is released, it can be implemented into `socket2`.

> I think WASIX is a no-go as that's vendor specific. Maybe as an alternative target, but I'm not how that would work at the moment.

I'm also not quite sure how a WASIX implementation should look, and to my knowledge, WASIX is only meant to be a polyfill until the official standard is fully developed.

> Overall I'm also ok with reducing the socket2 to something that is supported on wasi preview 1 (and only wasi) so that it can at least build.

I think it would also help when implementing `wasm32-wasip2`, if the imported functions work the same as in  WASI preview 2.

---

I would be really interested in giving this a try, but because of university, I don't have much free time. Do you know if there is some kind of tracking issue for `wasm32-wasip2` and the implementation of WASI preview 2 in rustc?

**Comment by Thomasdezeeuw** [comment]
> I think it would also help when implementing `wasm32-wasip2`, if the imported functions work the same as in WASI preview 2.
> 
> I would be really interested in giving this a try, but because of university, I don't have much free time. Do you know if there is some kind of tracking issue for `wasm32-wasip2` and the implementation of WASI preview 2 in rustc?

After a quick search I can't find any tracking issues, but let's focus on preview 1 initially. 

> Or some documentation on how to introduce a new target into `socket2`?

There aren't a lot of of docs, but I can give a quick overview.

The first thing would be to start running `cargo check --target wasm32-wasi` (or `wasm32-wasip1` in the future I suppose). For each failure add the appropriate `cfg` attributes, e.g. https://github.com/rust-lang/socket2/blob/6923954ce066e26a1d726ba2ca9d2b2c2176f3a2/src/lib.rs#L331 There will be a lot of these. But like I said I'm totally fine with just adding `not(wasm)` to most of the API to begin with.

Once `cargo check` works, try adding `--features all`, that enables even more API.

Once it builds/checks, ideally we find a way to run the tests. Since wasm can't create their own sockets (yet), this will be painful as most if not all of the tests do this. Feel free to skip this entirely.

Final step would be to add some kind of CI, again tests would be ideal, but running `cargo check` would be sufficient. That is (or should be) as easy as adding the wasm target to https://github.com/rust-lang/socket2/blob/6923954ce066e26a1d726ba2ca9d2b2c2176f3a2/.github/workflows/main.yml#L59,  https://github.com/rust-lang/socket2/blob/6923954ce066e26a1d726ba2ca9d2b2c2176f3a2/.github/workflows/main.yml#L97 and https://github.com/rust-lang/socket2/blob/6923954ce066e26a1d726ba2ca9d2b2c2176f3a2/Makefile#L3

**Comment by Thomasdezeeuw** [comment]
@LetsMelon if you have any questions feel free to open a draft pr and ping me. Or you can leave them here.

**Comment by LetsMelon** [comment]
@Thomasdezeeuw thanks for your quick reply and your help!

I will try to look into the codebase over the weekend. My current plan is to fix all `cfg`'s and copy all functions from `src/sys/unix.rs` into something like `src/sys/wasm.rs`, replacing their bodies with `todo!`. Maybe that will be enough for `socket2` to compile (but not work) with wasm.

**Comment by mirsella** [comment]
hello, if wasm sockets are still far, maybe in the mean time web_sys::websocket can be used, as most wasm usage is for web.
thats what the old wasm_sockets use.

personally it's because this crate is used in a bevy plugin, and i want to use bevy as a web app.

thanks !

**Comment by mirsella** [comment]
any update ?

**Comment by Thomasdezeeuw** [comment]
No update, nobody is working on this.
